core/src/main/java/org/infinispan/interceptors/impl/CacheWriterInterceptor.java

```diff
@@ -209,20 +209,28 @@ public Object visitReplaceCommand(InvocationContext ctx, ReplaceCommand command)
 
    @Override
    public Object visitPutMapCommand(InvocationContext ctx, PutMapCommand command) throws Throwable {
-      return invokeNextThenAccept(ctx, command, (rCtx, rCommand, rv) -> {
-         PutMapCommand putMapCommand = (PutMapCommand) rCommand;
-         if (!isStoreEnabled(putMapCommand) || rCtx.isInTxScope())
-            return;
+      return invokeNextThenAccept(ctx, command, handlePutMapCommandReturn);
+   }
 
-         Map<Object, Object> map = putMapCommand.getMap();
-         for (Object key : map.keySet()) {
-            if (isProperWriter(rCtx, putMapCommand, key)) {
-               storeEntry(rCtx, key, putMapCommand);
-            }
-         }
-         if (getStatisticsEnabled())
-            cacheStores.getAndAdd(map.size());
-      });
+   protected void handlePutMapCommandReturn(InvocationContext rCtx, VisitableCommand rCommand, Object rv) {
+      PutMapCommand putMapCommand = (PutMapCommand) rCommand;
+      if (!isStoreEnabled(putMapCommand) || rCtx.isInTxScope())
+         return;
+
+      processIterableBatch(rCtx, putMapCommand, BOTH, key -> !skipSharedStores(rCtx, key, putMapCommand));
+      processIterableBatch(rCtx, putMapCommand, PRIVATE, key -> skipSharedStores(rCtx, key, putMapCommand));
+   }
+
+   protected void processIterableBatch(InvocationContext ctx, PutMapCommand cmd, PersistenceManager.AccessMode mode, Predicate<Object> filter) {
+      Iterable<MarshalledEntry> iterable = () -> StreamSupport.stream(Spliterators.spliterator(cmd.getMap().keySet(), Spliterator.NONNULL), false)
+            .filter(filter)
+            .peek(Void -> {
+               if (getStatisticsEnabled())
+                  cacheStores.incrementAndGet();
```

Comment:
Could this just not be incremented by `cmd.getMap().size()` like we did before ?